Duplicati/Library/Utility/ReadLimitLengthStream.cs

```diff
@@ -138,14 +185,38 @@ public override void WriteByte(byte value)
 
         private int GetAllowedCount(int count)
         {
-            long pos = m_innerStream.Position;
-            long maxCount = m_endPosition - pos;
+            long pos = m_innerStream.Position - m_start;
+            long maxCount = m_length - pos;
+            if (pos < 0 || maxCount < 0)
+            {
+                // The stream is positioned before the starting point or after the end.
+                // Nothing can be read.
+                return 0;
+            }
+
             if (count > maxCount)
             {
                 return (int)maxCount;
             }
 
             return count;
         }
+
+        private long ClampInnerStreamPosition(long position)
+        {
+            // Note that this allows this stream to have positions in the range -1 to m_length.
+            // Reading at -1 or m_length should return nothing.

```

Comment:
This feels a bit strange, since other code might perform calculations using the `Position` property with the assumption that the value is nonnegative.  Would you be able to explain the reason for allowing `Position` to be negative?

Reply:
The reason I had done this was mostly an attempt at giving some consistent behavior when the underlying stream is positioned before the limited stream's starting point - I had originally been thinking that if the underlying stream was positioned before the start point, that Read() calls should fail (in part because if the stream couldn't be seeked, then we would be in a case where the stream reports a position of 0, but can't be read). However, I realized that even if we can't seek forward, we could at least read until reaching the start position, which removes that oddity. So I've revised this to clamp the reported position to 0 if the base stream is positioned before the start. Attempting to read will cause the stream to be seeked to start position and will then read from there.